faidx.c

```diff
@@ -53,6 +56,15 @@ DEALINGS IN THE SOFTWARE.
 #   define ABS(x) ((x)>=0?(x):-(x))
 #endif
 
+//new params required for output creation
+typedef struct output {
+    int isbgzip;                //is bgzip or uncompressed file
+    FILE *fp;                   //uncompressed file pointer
+    BGZF *bgzf_fp;              //bgzf file pointer
+    sam_global_args *gopt;      //options
+    kstring_t *buffer;
```

Comment:
This doesn't need to be a pointer with a calloc / free.  kstring_t is a fixed size struct and the memory allocation is handled internally, so you can just do `kstring_t buffer` instead of `*buffer` and then either `ks_initialise(&buffer);` in place of the calloc, or easier just change the initialisation of `struct output` to use `KS_INITIALIZE` in situ.  Eg:

```
struct output out = { 0, stdout, NULL, &ga, KS_INITIALIZE};
```

I tried it quickly (see ~jkb/work/samtools_master/samtools/faidx.c) and it's just a matter of &ks or ks.s instead of ks and ks->s in the various function calls.


That avoids callocs + the error check there and the free statement.

```
diff --git a/faidx.c b/faidx.c
index 652df65c..ddd376b9 100644
--- a/faidx.c
+++ b/faidx.c
@@ -62,7 +62,7 @@ typedef struct output {
     FILE *fp;                   //uncompressed file pointer
     BGZF *bgzf_fp;              //bgzf file pointer
     sam_global_args *gopt;      //options
-    kstring_t *buffer;
+    kstring_t buffer;
 } output;
 
 static unsigned char comp_base[256] = {
@@ -171,24 +171,17 @@ static int write_output(faidx_t *faid, output *out, const char *name, const int
     if (wrap_len <= 0)
         wrap_len = HTS_POS_MAX;
 
-    if (!out->buffer) {
-        if (!(out->buffer = calloc(1, sizeof(kstring_t)))) {
-            fprintf(stderr,"[faidx] Couldn't create buffer\n");
-            goto exit;
-        }
-    }
-
     seq = fai_fetch64(faid, name, &seq_len);
     if (rev && seq_len > 0) {
         reverse_complement(seq, seq_len);
     }
     //write the name
-    len = ksprintf(out->buffer, "%c%s%s\n", format == FAI_FASTA ? '>' : '@', name, rev ? neg_strand_name : pos_strand_name);
-    if (wrappedwrite(out, out->buffer->s, out->buffer->l) < len) {
+    len = ksprintf(&out->buffer, "%c%s%s\n", format == FAI_FASTA ? '>' : '@', name, rev ? neg_strand_name : pos_strand_name);
+    if (wrappedwrite(out, out->buffer.s, out->buffer.l) < len) {
         fprintf(stderr,"[faidx] Failed to write buffer\n");
         goto exit;
     }
-    ks_clear(out->buffer);
+    ks_clear(&out->buffer);
     //write bases
     if ((ret = write_line(faid, out, seq, name, ignore, wrap_len, seq_len) == EXIT_FAILURE)) {
         goto exit;
@@ -201,12 +194,12 @@ static int write_output(faidx_t *faid, output *out, const char *name, const int
             reverse(qual, seq_len);
         }
 
-        len = ksprintf(out->buffer, "+\n");
-        if (wrappedwrite(out, out->buffer->s, out->buffer->l) < len) {
+        len = ksprintf(&out->buffer, "+\n");
+        if (wrappedwrite(out, out->buffer.s, out->buffer.l) < len) {
             fprintf(stderr,"[faidx] Failed to write buffer\n");
             goto exit;
         }
-        ks_clear(out->buffer);
+        ks_clear(&out->buffer);
         if ((ret = write_line(faid, out, qual, name, ignore, wrap_len, seq_len) == EXIT_FAILURE)) {
             goto exit;
         }
@@ -297,7 +290,7 @@ int faidx_core(int argc, char *argv[], enum fai_format_options format)
     char *gzi_name = NULL; // specified compressed index name
     sam_global_args ga = SAM_GLOBAL_ARGS_INIT;
     int exit_status = EXIT_FAILURE, flushed = 0;
-    struct output out = { 0, stdout, NULL, &ga, NULL};          //data required for output writing
+    struct output out = { 0, stdout, NULL, &ga, KS_INITIALIZE};          //data required for output writing
     faidx_t *fai = NULL;
     hts_tpool *pool = NULL;
 
@@ -531,10 +524,7 @@ exit2:
         hts_tpool_destroy(pool);
     }
     sam_global_args_free(&ga);
-    if (out.buffer) {
-        ks_free(out.buffer);
-        free(out.buffer);
-    }
+    ks_free(&out.buffer);
     return exit_status;
 }
```

Reply:
Note there's nothing wrong with what you did!  The code works, but it's useful to learn easier ways of manipulating kstrings.

Reply:
updated.